### ISSUE TITLE
🌤🌤 CONTRIBUTE 🎉🎉

### ISSUE THREAD

**Issue by DefTruth** [DefTruth/CUDA-Learn-Notes#50]
<div align='left'>
  <img src='https://github.com/user-attachments/assets/b2578723-b7a7-4d8f-bcd1-5008947b808a' >
</div> 

## 🌤🌤目标

首先，任何kernel实现都欢迎，**本仓库学习/练习为主，性能最优非本仓库最终目标，先会用，然后再用好**。性能最优推荐直接使用cuBLAS, cuDNN, FlashAttention, TensorRT等官方实现。如果有感兴趣的kernel希望在本仓库实现，可以评论本issue（虽然我不一定有能力实现🌚），比如： 

#### ☕️☕️Kernel Trace
- [ ] xxx kernel 
- [ ] ...

## 👨‍💻👨‍💻代码规范 

提交代码需要遵循以下规范：

- 每个算子/kernel使用单独的目录，参考relu, gelu等
- 参考任一现有kernel，使用torch验证结果正确性
- 本仓库使用2空格作为缩进
- { }使用非上下对齐风格
- pragma unroll和当前for loop代码对齐
- 一行代码尽量不超过100字符
- 不使用/未ready的代码删除
- 其他想到再写🌚......

##  🎉🎉 致谢
感谢 @bear-zd, @wangzijian1010等为本仓库提供大量kernel实现 ~

## ☕️☕️Kernel Trace

- [x] swish kernel #85 by @wangzijian1010 
- [x] gelu kernel #66 by @bear-zd 
- [x] RoPE kernel #80 by @bear-zd 
- [x] pack elementwise_add
- [x] pack sigmoid
- [x] pack relu
- [x] histogram
- [x] warp/block reduce 
- [x] softmax
- [x] pack safe_softmax
- [x] pack layer-norm  
- [x] pack rms-norm 
- [x] flash-attn-1 f32
- [x] flash-attn-2 f16
- [ ] flash-attn-3 f8 ada
- [x] MMA(Tensor Cores) flash-attn-2 f16
- [x] warp segmv
- [x] warp hgemv 
- [x] bank confilcts reduce sgemm
- [x] pipeling sgemm
- [ ] split_k sgemm
- [x] pack LDST hgemm
- [x] bank confilcts reduce hgemm
- [x] pipeling hgemm
- [ ] split_k hgemm
- [x] cp.async hgemm
- [x] cp.async sgemm
- [x] WMMA API(Tensor Cores) hgemm
- [x] stage2/3/4 (Tensor Cores) hgemm
- [x] MMA PTX(Tensor Cores) hgemm
- [x] TF32 WMMA(Tensor Cores) sgemm
- [x] online_safe_softmax f32 #60 by @bear-zd 
- [x] pack online_safe_softmax #73 by @bear-zd 
- [x] embedding pack/unpack kernel #68 by @bear-zd 
- [x] mat transpose kernel #89 by @bear-zd 
- [x] Thread Block Swizzle hgemm
- [x] Thread Block Swizzle sgemm TF32
- [ ] batchnorm kernel
- [x] nms kernel #102 by @bear-zd 
- [x] HGEMM MMA Col Major 
- [ ] ... (any kernel)
